tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs

```diff
@@ -0,0 +1,367 @@
+        /// or (2) when the async method has logically ended.
+        /// In this phase we have the correct async context in hand because we already set it in the BeginMethod.
+        /// </summary>
+        /// <typeparam name="TTarget">Target object of the method. Note that it could be typeof(object) and not a concrete type</typeparam>
+        /// <typeparam name="TReturn">Return type</typeparam>
+        /// <param name="instance">Instance value</param>
+        /// <param name="returnValue">Return value</param>
+        /// <param name="exception">Exception value</param>
+        /// <param name="asyncState">Debugger asyncState</param>
+        /// <returns>LiveDebugger return structure</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static DebuggerReturn<TReturn> EndMethod_StartMarker<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, ref AsyncMethodDebuggerState asyncState)
+        {
+            if (!asyncState.IsActive)
+            {
+                return new DebuggerReturn<TReturn>(returnValue);
+            }
+
+            asyncState.SnapshotCreator.StartReturn();
+            asyncState.SnapshotCreator.CaptureInstance(asyncState.KickoffInvocationTarget, asyncState.KickoffInvocationTargetType);
+            if (exception != null)
+            {
+                asyncState.SnapshotCreator.CaptureException(exception);
+            }
+            else
+            {
+                asyncState.SnapshotCreator.CaptureLocal(returnValue, "@return", true);
+                asyncState.HasLocalsOrReturnValue = true;
+            }
+
+            EndMethodLogLocals(ref asyncState);
+            BeginMethodLogArgs(instance, ref asyncState);
+            return new DebuggerReturn<TReturn>(returnValue);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void EndMethodLogLocals(ref AsyncMethodDebuggerState asyncState)
+        {
+            // MethodMetadataInfo saves locals from MoveNext localVarSig,
+            // this isn't enough in async scenario because we need to extract more locals the may hoisted in the builder object
+            // and we need to subtract some locals that exist in the localVarSig but they are not belongs to the kickoff method
+            // For know we capturing here all locals the are hoisted (except known generated locals)
+            // and we capturing in LogLocal the locals form localVarSig
+            var kickOffMethodLocalsValues = asyncState.MethodMetadataInfo.AsyncMethodHoistedLocals;
+            for (var index = 0; index < kickOffMethodLocalsValues.Length; index++)
+            {
+                var local = kickOffMethodLocalsValues[index];
+                asyncState.SnapshotCreator.CaptureLocal(local.Field.GetValue(asyncState.MoveNextInvocationTarget), local.SanitizedName, index == 0 && !asyncState.HasLocalsOrReturnValue, local.Field.FieldType);
```

Comment:
Same as I noted before, we need to make sure `GetValue` won't fail us when we deal with (open/close) generics.